### PR TITLE
added min and max aggregators

### PR DESCRIPTION
**Pull request by jbloom22**

### CODE REVIEW

File: src/main/scala/is/hail/methods/Aggregators.scala [hunk 1]
```diff
@@ -371,6 +371,60 @@ class SumArrayAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T])
   def copy() = new SumArrayAggregator()
 }
 
+class MaxAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T]) extends TypedAggregator[Any] {
+
+  var _state: T = ev.zero
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Aggregators.scala]: private vars, not underscore

Comment (tpoterba) [inline src/main/scala/is/hail/methods/Aggregators.scala]: prefer `private var state: T = _`

File: src/main/scala/is/hail/methods/Aggregators.scala [hunk 2]
```diff
@@ -371,6 +371,60 @@ class SumArrayAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T])
...
+
+  def combOp(agg2: this.type) {
+    if (_isNull || !agg2._isNull && ev.compare(_state, agg2._state) > 0)
+          _state = agg2._state
+  }
+
+  def copy() = new MaxAggregator()
+}
+
+class MinAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T]) extends TypedAggregator[Any] {
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Aggregators.scala]: If you were worried about boxing the value at any point, this definitely does that.

File: src/main/scala/is/hail/methods/Aggregators.scala [hunk 3]
```diff
@@ -371,6 +371,60 @@ class SumArrayAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T])
...
+
+  var _state: T = ev.zero
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Aggregators.scala]: see above

File: src/main/scala/is/hail/expr/HailRep.scala [hunk 4]
```diff
@@ -3,7 +3,7 @@ package is.hail.expr
 import is.hail.utils.Interval
 import is.hail.variant.{AltAllele, Call, Genotype, Locus, Variant}
 
-trait HailRep[T] { self =>
+trait HailRep[+T] { self =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/HailRep.scala]: I'm not sure whether it's better to do this or to have an `AnyHr(typ: Type) extends HailRep[Any]` class

Comment (cseed) [inline src/main/scala/is/hail/expr/HailRep.scala]: I don't think this is right.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 5]
```diff
@@ -1595,6 +1595,24 @@ object FunctionRegistry {
...
+  registerAggregator[Double, Any]("max", () => new MaxAggregator[Double](), maxAggDocstring)(aggregableHr(doubleHr), boxeddoubleHr)
+
+  val minAggDocstring = """Compute the minimum of all non-missing elements. The empty min is missing."""
+  registerAggregator[Int, Any]("min", () => new MinAggregator[Int]()(ev = implicitly[Numeric[Int]], ct = ClassTag.apply(classOf[Int])), minAggDocstring)(aggregableHr(intHr), boxedintHr)
+
+  registerAggregator[Long, Any]("min", () => new MinAggregator[Long](), minAggDocstring)(aggregableHr(longHr), boxedlongHr)
+
+  registerAggregator[Float, Any]("min", () => new MinAggregator[Float](), minAggDocstring)(aggregableHr(floatHr), boxedfloatHr)
+
+  registerAggregator[Double, Any]("min", () => new MinAggregator[Double](), minAggDocstring)(aggregableHr(doubleHr), boxeddoubleHr)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: This should be `[Double, java.lang.Double]`, that will fix the `HailRep[+T]` issue.

Comment (cseed) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: I think these should be `MinAggregator[java.lang.Double]`, etc.  Otherwise, null gets converted to 0.0.  Add a test case that verifies that min/max aggregation of an empty/null values returns null.

File: src/main/scala/is/hail/methods/Aggregators.scala [hunk 6]
```diff
@@ -371,6 +371,60 @@ class SumArrayAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T])
...
+        _state = x.asInstanceOf[T]
+        _isNull = false
+      } else
+        if (ev.compare(x.asInstanceOf[T], _state) > 0)
+          _state = x.asInstanceOf[T]
+    }
+  }
+
+  def combOp(agg2: this.type) {
+    if (_isNull || !agg2._isNull && ev.compare(_state, agg2._state) > 0)
```
Comment (cseed) [inline src/main/scala/is/hail/methods/Aggregators.scala]: Prefer parentheses mixing && and ||.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 7]
```diff
@@ -1577,7 +1577,7 @@ object FunctionRegistry {
 
     **Examples**
 
-    Compute the counts of each allele per variant:
+    Count the total number of occurances of each allele across samples, per variant:
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: spelling

File: src/test/scala/is/hail/methods/AggregatorSuite.scala [hunk 8]
```diff
@@ -123,6 +123,38 @@ class AggregatorSuite extends SparkSuite {
     p.check()
   }
 
+  @Test def testMax() {
+    val p = Prop.forAll(VariantSampleMatrix.gen(hc, VSMSubgen.random)) { vds =>
```
Comment (cseed) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: I would say cooking up random VSMs is way overkill to test min/max.  Just test it on some small KeyTables (with nulls, etc.)

File: src/main/scala/is/hail/methods/Aggregators.scala (hunk 1 above)
Comment (jbloom22) [inline src/main/scala/is/hail/methods/Aggregators.scala]: should be separate PR to change throughout

**Comment by jbloom22** [comment]
Back to @cseed and @tpoterba .

Tim, Cotton agrees that changing convention from _ to private should be separate PR on all functions here.

File: src/main/scala/is/hail/utils/richUtils/Implicits.scala [hunk 9]
```diff
@@ -98,3 +99,106 @@ trait Implicits {
 
   implicit def toRichJValue(jv: JValue): RichJValue = new RichJValue(jv)
 }
+
+object NumericJavaLang {
```
Comment (tpoterba) [inline src/main/scala/is/hail/utils/richUtils/Implicits.scala]: this is the wrong place for this object, I think. Put this in a standalone class in the richUtils package and add `implicit val`s in `Implicits.scala`

File: src/test/scala/is/hail/methods/AggregatorSuite.scala [hunk 10]
```diff
@@ -123,6 +126,24 @@ class AggregatorSuite extends SparkSuite {
     p.check()
   }
 
+  @Test def testMaxMin() {
+    val rdd = sc.parallelize(Seq(
+      Annotation("a",  0,    1, -1.0, null,  1,   -1, null, null),
```
Comment (cseed) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: Why use 1.0 and 1f?  Seems like 1-4 are the same as 5-8.  Why not just compute the min and max of 1-4?

File: src/main/scala/is/hail/utils/NumericPair.scala [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package is.hail.utils
+
+import scala.math.Numeric.{DoubleIsFractional, FloatIsFractional, IntIsIntegral, LongIsIntegral}
```
Comment (cseed) [inline src/main/scala/is/hail/utils/NumericPair.scala]: No longer needed.

File: src/main/scala/is/hail/utils/richUtils/Implicits.scala [hunk 12]
```diff
@@ -13,6 +13,7 @@ import org.json4s.JValue
 
 import scala.collection.{TraversableOnce, mutable}
 import scala.language.implicitConversions
+import scala.math.Ordering
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/Implicits.scala]: Is this necessary?  There are no other changes to the file.

File: src/main/scala/is/hail/utils/NumericPair.scala [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+trait NumericPair[T, BoxedT >: Null] extends Serializable {
+
+  def box(t: T): BoxedT
+
+  def unbox(u: BoxedT): T
+
+  implicit val numeric: scala.math.Numeric[T]
+}
+
+object NumericPairs {
```
Comment (cseed) [inline src/main/scala/is/hail/utils/NumericPair.scala]: I'd make this a trait and have utils package object inherit from it.  Probably call it NumericPairImplicits or something.

**Comment by cseed** [comment]
I shudder to think about the crazy boxing that's happening in something like `def unbox(u: BoxedT): T` but we can deal with that later.  Doing `@specialized(Int, Long, Float, Double)` in NumericPair and Min/MaxAggregator would probably do the trick but someone probably needs to look through the generated bytecode and convince themselves it's doing the right thing.  I'm still a little skeptical about @specialized after the compiler bug interacting with private[this].

**Comment by jbloom22** [comment]
@tpoterba it's in your court.

File: src/main/scala/is/hail/methods/Aggregators.scala [hunk 14]
```diff
@@ -371,6 +371,50 @@ class SumArrayAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T])
   def copy() = new SumArrayAggregator()
 }
 
+class MaxAggregator[T, BoxedT >: Null](implicit ev: NumericPair[T, BoxedT], ct: ClassTag[T]) extends TypedAggregator[BoxedT] {
+
+  import ev.numeric
+  import Ordering.Implicits._
+
+  var _state: BoxedT = _
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Aggregators.scala]: why can't _state be unboxed if you've got both the primitive and boxed types hanging around?

Comment (jbloom22) [inline src/main/scala/is/hail/methods/Aggregators.scala]: I've just pushed a version that takes this approach, and I like it better too. You?

File: src/main/scala/is/hail/methods/Aggregators.scala [hunk 15]
```diff
@@ -371,6 +371,56 @@ class SumArrayAggregator[T](implicit ev: scala.math.Numeric[T], ct: ClassTag[T])
...
+  def seqOp(x: Any) {
+    if (x != null && (_isNull || ev.unbox(x.asInstanceOf[BoxedT]) < _state)) {
+      _isNull = false
+      _state = ev.unbox(x.asInstanceOf[BoxedT])
+    }
+  }
+
+  def combOp(agg2: this.type) {
+    if (!agg2._isNull && (_isNull || agg2._state < _state))
+      _state = agg2._state
```
Comment (cseed) [inline src/main/scala/is/hail/methods/Aggregators.scala]: You need to update _isNull here.  Your tests didn't catch this since they use only a single partition.

**Comment by jbloom22** [comment]
back to you

File: src/test/scala/is/hail/methods/AggregatorSuite.scala [hunk 16]
```diff
@@ -123,6 +126,31 @@ class AggregatorSuite extends SparkSuite {
...
+  @Test def testMaxMin() {
+    val rdd1 = sc.parallelize(Seq(
+      Annotation("a",  0, null, -1, -1.0, null),
+      Annotation("a", -1, null, -2, 1.0, null)), numSlices = 1)
+
+    val rdd2 = sc.parallelize(Seq(
+      Annotation("a",  1, -1, -3, null, null),
+      Annotation("a",  0, -2, -4, 0.0, null)), numSlices = 1)
+
+    val rdd = rdd1.union(rdd2)
```
Comment (cseed) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: Why not just use one parallelize and numSlices = 2?

Comment (jbloom22) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: because I wanted to make sure I had two elements in each of the partitions

Comment (cseed) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: Why do you need two per partition?  You just need to test each side being null.  Why not:

null, null, 1
1, null, null

parallelize splits evenly:

```
scala> sc.parallelize(Seq(1, 2, 3, 4), 2).mapPartitions { it => Iterator(it.length) }.collect()
res1: Array[Int] = Array(2, 2)
```

Comment (jbloom22) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: Two per partition because I wanted to test null behavior inside both combOp and seqOp. I'm happy to change to one RDD, but I felt this approach made it transparent in code what was happening.

Comment (cseed) [inline src/test/scala/is/hail/methods/AggregatorSuite.scala]: I guess I don't know what you mean by "the null behavior" and how this achieves that.

I guess you'd want to test a combOp of a non-null and a null, but you don't do that.  Or a seqOp of a non-null followed by a null, but you don't do that either.

Yes, I think the merge is less clear than than sc.parallelize(..., 2).

**Comment by jbloom22** [comment]
added space. back to you
